### DISCUSSION TITLE
UPGRADINGPRODUCTCODE is not being set in the uninstalling setup

### DISCUSSION THREAD

**Discussion by maqsoodmushtaq** [wixtoolset/issues#7087] [Questions]
**MajorUpgrade** element is scheduled after install finalize in our product's MSI:

```
    <MajorUpgrade Schedule="afterInstallFinalize" AllowSameVersionUpgrades="yes" DowngradeErrorMessage="!(loc.NewerVersionInstalled)" IgnoreRemoveFailure="no"/>
```
There are some folders written by app at runtime that we want to keep on upgrade and only remove on uninstall initiated from Add/Remove programs. So we use this condition: (NOT UPGRADINGPRODUCTCODE) AND (REMOVE="ALL").

```
        <DirectoryRef Id="TARGETDIR">
		...		
            <Directory Id="LocalAppDataFolder"
                xmlns:util="http://schemas.microsoft.com/wix/UtilExtension">
                <Component Id="CreatePrivateMyAppFolder" Guid="FA1F4375-71DA-4E61-9A02-BE7FD2D4C87D">
                    <RegistryValue Root="HKCU" Key="Software\Company\Product" Name="PrivateFolderMyApp" Type="string" Value="[PrivateDataMyApp]" KeyPath="yes"/>
                </Component>
                <Component Id="RemoveLocalAppDataMyAppUninstall" Guid="*" Transitive="yes">
                    <Condition><![CDATA[(NOT UPGRADINGPRODUCTCODE) AND (REMOVE="ALL")]]></Condition>
                    <RegistryValue Root="HKCU" Key="Software\Company\Product" Name="PrivateFolderMyApp" Type="string" Value="[PrivateDataMyApp]" KeyPath="yes"/>
                    <util:RemoveFolderEx On="uninstall" Property="PRIVATEMYAPPFOLDER"/>
                </Component>
		<Directory></Directory>
		...
	    </Directory>
	</DirectoryRef>
```

I need to change the MajorUpgrade schedule from **_afterInstallFinalize_** to **_afterInstallInitialize_** for some new requirements. However when I run new MSI with **_afterInstallInitialize_**, folders written by runtime are being removed which indicates to me that the condition UPGRADINGPRODUCTCODE is not being set in the uninstalling setup.

Any comments to explain/debug this will be greatly appreciated.

**Comment by robmen** [comment]
`MajorUpgrade` element sets `WIX_UPGRADE_DETECTED` not `UPGRADINGPRODUCTCODE` as per the documentation:  https://wixtoolset.org/docs/v3/xsd/wix/majorupgrade/

**Reply by maqsoodmushtaq** [reply]
`MajorUpgrade` element schedules the `RemoveExistingProducts` action based on Schedule attribute (Schedule="afterInstallFinalize") which sets the `UPGRADINGPRODUCTCODE` property. That is why I am using it in my condition. I tried `NOT WIX_UPGRADE_DETECTED` but did not work either.

**Reply by robmen** [reply]
Sorry, I skimmed over your question too quickly. You are in the nested uninstall, not the parent package.

A verbose log file of the upgrade should show you the expected action states of all the components for both the new and old MSIs and will show you the state of `UPGRADINGPRODUCTCODE` as the install progresses. There will be lots of data in the log file.

**Comment by maqsoodmushtaq** [comment]
From the logs, both `UPGRADINGPRODUCTCODE` and `REMOVE` properties are set. Based on that the condition `(NOT UPGRADINGPRODUCTCODE) AND (REMOVE="ALL")` should evaluate to false for the component `RemoveLocalAppDataMyAppUninstall`?

```
MSI (s) (C4:58) [22:58:11:060]: Doing action: RemoveExistingProducts
Action 22:58:11: RemoveExistingProducts. Removing applications
Action start 22:58:11: RemoveExistingProducts.
RemoveExistingProducts: Application: {8F890AE0-BE0A-5ED9-B406-F7459B3390F9}, Command line: UPGRADINGPRODUCTCODE={70705091-36C8-5619-9E35-73E455CA17F7} CLIENTPROCESSID=4756 CLIENTUILEVEL=0 REMOVE=ALL
....
MSI (s) (C4:4C) [22:58:11:076]: Command Line: UPGRADINGPRODUCTCODE={70705091-36C8-5619-9E35-73E455CA17F7} CLIENTPROCESSID=4756 CLIENTUILEVEL=0 REMOVE=ALL 
MSI (s) (C4:4C) [22:58:11:279]: Dir (target): Key: _PRIVATEMYAPPFOLDER_4	, Object: C:\Users\Windows_10\AppData\Local\MyApp\
MSI (s) (C4:4C) [22:58:11:279]: Dir (target): Key: _PRIVATEMYAPPFOLDER_3	, Object: C:\Users\Windows_10\AppData\Local\MyApp\1753de9b-15a7-49b1-8715-f93a967d12e5\
...
MSI (s) (C4:4C) [22:58:11:826]: Doing action: InstallValidate
MSI (s) (C4:4C) [22:58:11:826]: Component: RemoveLocalAppDataMyAppUninstall; Installed: Local;   Request: Absent;   Action: Absent
...
MSI (s) (C4:4C) [22:58:12:732]: Doing action: RemoveFiles
MSI (s) (C4:4C) [22:58:12:919]: Counted 6 foreign folders to be removed.
MSI (s) (C4:4C) [22:58:12:919]: Removing foreign folder: C:\Users\Windows_10\AppData\Local\MyApp\1753de9b-15a7-49b1-8715-f93a967d12e5\
MSI (s) (C4:4C) [22:58:12:919]: Removing foreign folder: C:\Users\Windows_10\AppData\Local\MyApp\
```

**Reply by barnson** [reply]
Component conditions only affect install and, with the transitive bit set, reinstall. Uninstall isn't affected. `RemoveFolderEx` in WiX v4 has a `Condition` that lets you do what you want to do.

**Reply by maqsoodmushtaq** [reply]
Thanks @barnson - this helped me to move in right direction to explore different solution.
